### PR TITLE
[otbn] add modexp subroutine for pubexp 65537

### PR DESCRIPTION
**Pull request by felixmiller**
This adds the last subroutine to have both RSA encrypt/verify + decrypt/sign self-contained in modexp.

### CODE REVIEW

File: sw/otbn/code-snippets/modexp.S [hunk 1]
```diff
@@ -1288,6 +1288,142 @@ modexp:
...
+    /* store selected limb to dmem */
+    bn.sid    x8, 0(x21++)
+    addi      x8, x8, 1
+
+  /* 65537 = 0b10000000000000001
+                ^<< 16 x sqr >>^   */
+  loopi      16, 2
+    /* square: out = montmul(out, out) */
+    jal       x1, sqrx_exp
+    nop
```
Comment (rswarbrick) [inline sw/otbn/code-snippets/modexp.S:1402]: This "jump inside a loop" precludes loop implementations based on counting instructions, which I think we were considering at one point. @GregAC: is this ok with your planned implementation?

Also, I wonder whether we can avoid needing the `nop` (does the loop match check look at link addresses?). If not, we'll have to document that quite carefully.

Comment (felixmiller) [reply sw/otbn/code-snippets/modexp.S:1402]: I am afraid the jump-in-loop feature is sth. we really need. We are using this a lot for multiplications and the Montgomery loop.

Regarding `nop`: I thought our conclusion from #3609 was a bit that we need it. But I agree we should definitely document this and probably have the tool chain throw a warning if last instruction in a loop body is a `JAL`

Comment (GregAC) [reply sw/otbn/code-snippets/modexp.S:1402]: My current RTL will work with branches/jumps in loops provided they're not the final instruction of a loop body.

Branches/jumps in a loop aren't an inherent problem, just opens the door to weird behaviour the hardware will have a hard time detecting. As it's all hand-crafted assembly we're probably fine with odd behaviour happening if you do something weird/stupid (e.g. branch from loop body then never branch back or branch back above the and start it again), indeed we could probably come up with some kind of static analysis tool to detect the dodgy cases and/or add some checking in the simulator for it.

The real question is around whether this is a security concern and whether we'd want hardware to detect odd behaviour and halt with an error.

I've got a big list of loop corner cases to go over, so let's maybe hold this discussion until I've got that out.

**Review by imphil — APPROVED** [review]
LGTM! With the nop as last instruction in the loop @GregAC's RTL should be able to cope (from what I understand), so we can get this in and modify as needed later on.

**Comment by felixmiller** [comment]
Added another commit to change the binding of the symbols for the `modexp` and `modexp_65537` routines.

**Review by imphil — APPROVED** [review]
LGTM. @rswarbrick on the global symbols: they look right, even though they don't fully solve the problem we had for now. OK to get this in, or should we remove the second commit?

**Comment by rswarbrick** [comment]
Yep, they look good. I could have sworn that I managed a multi-object link without `.globl` lines in the past, but I don't think that can be right (since assembly symbols are local by default), so I'm clearly going mad :-)

**Pull request merged**